File: day-13/shuttle_search.py
import math

input_file = 'day-13/input.txt'

def part1(input):
  timestamp, bus_ids = int(input.split('\n')[0]), input.split('\n')[1]
  bus_ids = [int(id) for id in bus_ids.split(',') if id != 'x']
  print(timestamp, bus_ids)
  best_next_bus = timestamp # replace with big int
  bus_id = -1
  for id in bus_ids:
    next_bus = id - (timestamp % id)
    print(id, next_bus)
    if next_bus < best_next_bus:
      best_next_bus = next_bus
      bus_id = id

  return bus_id * best_next_bus

# Based on https://rosettacode.org/wiki/Chinese_remainder_theorem#Python
def chinese_remainder(n, a):
    N = math.prod(n)
    sum = 0
    for n_i, a_i in zip(n, a):
        p = N // n_i
        sum += a_i * inverse_mod(p, n_i) * p
    return sum % N

def inverse_mod(product, n):
  return pow(product, -1, n)

def part2(input):
  bus_ids = input.split('\n')[1].split(',')
  n = []
  a = []

  for index, id in enumerate(bus_ids):
    if id != 'x':
      id = int(id)
      n.append(id)
      a.append(-index % id)
      # schedule[index] = id
  # print(schedule)

  return chinese_remainder(n, a)

if __name__ == "__main__":
  with open(input_file) as f:
    data = f.read()
  print("Part 1: ", part1(data))
  print("Part 2: ", part2(data))

File: day-13/shuttle_search.go
package main

import (
	"fmt"
	"io/ioutil"
	"math"
	"strconv"
	"strings"
)

const input_file string = "./day-13/input.txt"

func ReadInput(input string) (int, map[int]int) {
	lines := strings.Split(strings.TrimSpace(input), "\n")
	timestamp, _ := strconv.Atoi(lines[0])
	schedule := make(map[int]int)
	for i, id := range strings.Split(strings.TrimSpace(lines[1]), ",") {
		if id != "x" {
			bus_id, _ := strconv.Atoi(id)
			schedule[i] = bus_id
		}
	}

	return timestamp, schedule
}

func Part1(input string) int {
	timestamp, schedule := ReadInput(input)

	next_bus_time := math.MaxInt64
	bus_id := -1
	for _, id := range schedule {
		next_bus := id - (timestamp % id)
		if next_bus < next_bus_time {
			next_bus_time = next_bus
			bus_id = id
		}
	}
	return bus_id * next_bus_time
}

// Based on comment https://www.reddit.com/r/adventofcode/comments/kc4njx/2020_day_13_solutions/gfo9bgb/
func Part2(input string) int {
	_, schedule := ReadInput(input)
	time, lcm := 0, 1
	for i, bus_id := range schedule {
		for (time+i)%bus_id != 0 {
			time += lcm
		}
		lcm *= bus_id
	}
	return time
}

func main() {
	b, err := ioutil.ReadFile(input_file)
	if err != nil {
		fmt.Print(err)
	}
	data := string(b)

	fmt.Printf("Part 1: %d\n", Part1(data))
	fmt.Printf("Part 2: %d\n", Part2(data))
}
